### PR TITLE
server: add workers started stat

### PR DESCRIPTION
**Pull request by ramaraochavali**
Description: Adds a stat for `workers_started` that indicates listeners have been initialized on workers. With out this stat, we have to manually ping the listeners in readiness probes to check if they are initialized. See https://github.com/envoyproxy/envoy/issues/8321 for more conversations on this.
Risk Level: Low
Testing: Updated tests
Docs Changes: Updated
Release Notes: Updated

### CODE REVIEW

File: source/server/listener_manager_impl.cc [hunk 1]
```diff
@@ -669,18 +673,28 @@ bool ListenerManagerImpl::removeListener(const std::string& name) {
...
   uint32_t i = 0;
+  const auto listeners_pending_init =
+      std::make_shared<std::atomic<uint64_t>>(workers_.size() * active_listeners_.size());
   for (const auto& worker : workers_) {
     ASSERT(warming_listeners_.empty());
     for (const auto& listener : active_listeners_) {
-      addListenerToWorker(*worker, *listener);
+      addListenerToWorker(*worker, *listener, [this, listeners_pending_init]() {
+        if (--(*listeners_pending_init) == 0) {
+          workers_started_ = true;
```
Comment (ramaraochavali) [inline source/server/listener_manager_impl.cc]: I have moved `workers_started_` variable also here to truly reflect its status. If there are any objections/risks, I can revert that change.

Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: Looks `workers_started_` has some side affects, especially when trying to `stopWorkers` immediately before it is actually set to true - so reverted this change.

File: test/server/listener_manager_impl_test.cc [hunk 2]
```diff
@@ -749,12 +757,28 @@ version_info: version2
...
+  // Validate that workers_started stat is zero before calling startWorkers.
+  EXPECT_EQ(0, server_.stats_store_
+                   .gauge("listener_manager.workers_started", Stats::Gauge::ImportMode::NeverImport)
+                   .value());
+
   // Start workers.
   EXPECT_CALL(*worker_, addListener(_, _));
   EXPECT_CALL(*worker_, start(_));
   manager_->startWorkers(guard_dog_);
+  // Validate that workers_started stat is still zero before workers intimate the status via
```
Comment (zuercher) [inline test/server/listener_manager_impl_test.cc]: nit: intimate should be initialize

Comment (ramaraochavali) [reply test/server/listener_manager_impl_test.cc]: @zuercher Thanks for the review. I meant actually "initimate" only here -  I meant to say when workers initimate the listener initialization status via completion call back - then only this stat is updated. If It is not clear, I can update or else please suggest the better wording here. I will update

Comment (zuercher) [reply test/server/listener_manager_impl_test.cc]: Maybe just "set" then?

Comment (ramaraochavali) [reply test/server/listener_manager_impl_test.cc]: sure. Changed it. PTAL

File: source/server/listener_manager_impl.cc [hunk 3]
```diff
@@ -686,17 +690,26 @@ void ListenerManagerImpl::startWorkers(GuardDog& guard_dog) {
   ASSERT(!workers_started_);
   workers_started_ = true;
   uint32_t i = 0;
+  const auto listeners_pending_init =
+      std::make_shared<std::atomic<uint64_t>>(workers_.size() * active_listeners_.size());
```
Comment (mattklein123) [inline source/server/listener_manager_impl.cc]: nit: you could simplify this by just using a `Cleanup` object which in its destructor sets the stat. This would work as it's captured by all the lambdas. See https://github.com/envoyproxy/envoy/blob/daf4751a080802e5b0f87f274d45650acab0ec74/source/server/server.cc#L644-L645.

This seems to be the pattern we are using now for this kind of thing. cc @jmarantz @eziskind

Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: OK. Changed. PTAL.

File: source/server/listener_manager_impl.cc [hunk 4]
```diff
@@ -686,17 +691,22 @@ void ListenerManagerImpl::startWorkers(GuardDog& guard_dog) {
   ASSERT(!workers_started_);
   workers_started_ = true;
   uint32_t i = 0;
+
+  std::shared_ptr<void> workers_started(new Cleanup([this]() { stats_.workers_started_.set(1); }));
```
Comment (jmarantz) [inline source/server/listener_manager_impl.cc]: why not write it like this:
```
auto workers_started = std::make_shared<Cleanup>[this]({ stats_.workers_started_.set(1); });
```

Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: I am fine with that. Will change

Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: Changed. PTAL

File: source/server/listener_manager_impl.cc (hunk 3 above)
Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: That change is causing ASAN failure like 
`2020-01-10T05:14:54.1963895Z SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior source/server/listener_manager_impl.cc:695:61 in 
`
See the logs here https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/26615/logs/62

So I reverted it - not sure how to fix it. If you have any ideas let me know, I can look. Otherwise, let us go with this for now.

Comment (mattklein123) [reply source/server/listener_manager_impl.cc]: That link is not loading for me for some reason. Can you paste the ASAN error here?

/wait-any

Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: hmm.. not sure why it is loading. I will put back the cleanup change to get the logs.

Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: Here is log 
```
2020-01-11T03:45:23.3567727Z [2020-01-11 03:45:15.999][70][info][main] [source/server/server.cc:595] exiting
2020-01-11T03:45:23.3568335Z source/server/listener_manager_impl.cc:695:61: runtime error: member access within address 0x6120000676c0 which does not point to an object of type 'Envoy::Server::ListenerManagerImpl'
2020-01-11T03:45:23.3568600Z 0x6120000676c0: note: object has invalid vptr
2020-01-11T03:45:23.3568794Z  50 01 00 28  55 01 80 3f 00 00 00 00  00 29 02 00 20 62 00 00  58 2a 02 00 20 62 00 00  d8 76 06 00
2020-01-11T03:45:23.3568978Z               ^~~~~~~~~~~~~~~~~~~~~~~
2020-01-11T03:45:23.3569104Z               invalid vptr
2020-01-11T03:45:23.3569312Z     #0 0x1daa18a8 in Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_5::operator()() const /proc/self/cwd/source/server/listener_manager_impl.cc:695:61
2020-01-11T03:45:23.3569702Z     #1 0x1daa1761 in decltype(std::__1::forward<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_5&>(fp)()) std::__1::__invoke<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_5&>(Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_5&) /opt/llvm/bin/../include/c++/v1/type_traits:3530:1
2020-01-11T03:45:23.3570096Z     #2 0x1daa15a1 in void std::__1::__invoke_void_return_wrapper<void>::__call<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_5&>(Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_5&) /opt/llvm/bin/../include/c++/v1/__functional_base:348:9
2020-01-11T03:45:23.3570442Z     #3 0x1daa14ef in std::__1::__function::__alloc_func<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_5, std::__1::allocator<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_5>, void ()>::operator()() /opt/llvm/bin/../include/c++/v1/functional:1533:16
2020-01-11T03:45:23.3570795Z     #4 0x1da9c6f4 in std::__1::__function::__func<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_5, std::__1::allocator<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_5>, void ()>::operator()() /opt/llvm/bin/../include/c++/v1/functional:1707:12
2020-01-11T03:45:23.3571081Z     #5 0x14c1d139 in std::__1::__function::__value_func<void ()>::operator()() const /opt/llvm/bin/../include/c++/v1/functional:1860:16
2020-01-11T03:45:23.3571297Z     #6 0x14c1c09a in std::__1::function<void ()>::operator()() const /opt/llvm/bin/../include/c++/v1/functional:2419:12
2020-01-11T03:45:23.3571886Z     #7 0x1584b03a in Envoy::Cleanup::~Cleanup() /proc/self/cwd/bazel-out/k8-dbg/bin/source/common/common/_virtual_includes/cleanup_lib/common/common/cleanup.h:14:16
2020-01-11T03:45:23.3572172Z     #8 0x1db491d9 in std::__1::__shared_ptr_emplace<Envoy::Cleanup, std::__1::allocator<Envoy::Cleanup> >::__on_zero_shared() /opt/llvm/bin/../include/c++/v1/memory:3581:23
2020-01-11T03:45:23.3572508Z     #9 0x148b1af6 in std::__1::__shared_count::__release_shared() /opt/llvm/bin/../include/c++/v1/memory:3415:9
2020-01-11T03:45:23.3572722Z     #10 0x148b185d in std::__1::__shared_weak_count::__release_shared() /opt/llvm/bin/../include/c++/v1/memory:3457:27
2020-01-11T03:45:23.3572937Z     #11 0x1dacf881 in std::__1::shared_ptr<Envoy::Cleanup>::~shared_ptr() /opt/llvm/bin/../include/c++/v1/memory:4393:19
2020-01-11T03:45:23.3573164Z     #12 0x1da56a84 in Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_6::~$_6() /proc/self/cwd/source/server/listener_manager_impl.cc:700:47
2020-01-11T03:45:23.3573420Z     #13 0x1daa4ea4 in std::__1::__compressed_pair_elem<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_6, 0, false>::~__compressed_pair_elem() /opt/llvm/bin/../include/c++/v1/memory:2133:8
2020-01-11T03:45:23.3573728Z     #14 0x1daa53f7 in std::__1::__compressed_pair<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_6, std::__1::allocator<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_6> >::~__compressed_pair() /opt/llvm/bin/../include/c++/v1/memory:2210:7
2020-01-11T03:45:23.3574063Z     #15 0x1daa92aa in std::__1::__function::__alloc_func<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_6, std::__1::allocator<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_6>, void ()>::destroy() /opt/llvm/bin/../include/c++/v1/functional:1552:38
2020-01-11T03:45:23.3574404Z     #16 0x1daa33f4 in std::__1::__function::__func<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_6, std::__1::allocator<Envoy::Server::ListenerManagerImpl::startWorkers(Envoy::Server::GuardDog&)::$_6>, void ()>::destroy() /opt/llvm/bin/../include/c++/v1/functional:1689:10
2020-01-11T03:45:23.3574695Z     #17 0x149c8a7f in std::__1::__function::__value_func<void ()>::~__value_func() /opt/llvm/bin/../include/c++/v1/functional:1819:19
2020-01-11T03:45:23.3574919Z     #18 0x149bf044 in std::__1::function<void ()>::~function() /opt/llvm/bin/../include/c++/v1/functional:2406:43
2020-01-11T03:45:23.3575708Z     #19 0x1da8f07d in Envoy::Server::ListenerManagerImpl::addListenerToWorker(Envoy::Server::Worker&, Envoy::Server::ListenerImpl&, std::__1::function<void ()>)::$_4::operator()(bool) const::'lambda'()::~() /proc/self/cwd/source/server/listener_manager_impl.cc:592:31
2020-01-11T03:45:23.3576557Z     #20 0x1da93444 in std::__1::__compressed_pair_elem<Envoy::Server::ListenerManagerImpl::addListenerToWorker(Envoy::Server::Worker&, Envoy::Server::ListenerImpl&, std::__1::function<void ()>)::$_4::operator()(bool) const::'lambda'(), 0, false>::~__compressed_pair_elem() /opt/llvm/bin/../include/c++/v1/memory:2133:8
2020-01-11T03:45:23.3577637Z     #21 0x1da939d7 in std::__1::__compressed_pair<Envoy::Server::ListenerManagerImpl::addListenerToWorker(Envoy::Server::Worker&, Envoy::Server::ListenerImpl&, std::__1::function<void ()>)::$_4::operator()(bool) const::'lambda'(), std::__1::allocator<Envoy::Server::ListenerManagerImpl::addListenerToWorker(Envoy::Server::Worker&, Envoy::Server::ListenerImpl&, std::__1::function<void ()>)::$_4::operator()(bool) const::'lambda'()> >::~__compressed_pair() /opt/llvm/bin/../include/c++/v1/memory:2210:7
2020-01-11T03:45:23.3578772Z     #22 0x1da9706a in std::__1::__function::__alloc_func<Envoy::Server::ListenerManagerImpl::addListenerToWorker(Envoy::Server::Worker&, Envoy::Server::ListenerImpl&, std::__1::function<void ()>)::$_4::operator()(bool) const::'lambda'(), std::__1::allocator<Envoy::Server::ListenerManagerImpl::addListenerToWorker(Envoy::Server::Worker&, Envoy::Server::ListenerImpl&, std::__1::function<void ()>)::$_4::operator()(bool) const::'lambda'()>, void ()>::destroy() /opt/llvm/bin/../include/c++/v1/functional:1552:38
2020-01-11T03:45:23.3579915Z     #23 0x1da91c98 in std::__1::__function::__func<Envoy::Server::ListenerManagerImpl::addListenerToWorker(Envoy::Server::Worker&, Envoy::Server::ListenerImpl&, std::__1::function<void ()>)::$_4::operator()(bool) const::'lambda'(), std::__1::allocator<Envoy::Server::ListenerManagerImpl::addListenerToWorker(Envoy::Server::Worker&, Envoy::Server::ListenerImpl&, std::__1::function<void ()>)::$_4::operator()(bool) const::'lambda'()>, void ()>::destroy_deallocate() /opt/llvm/bin/../include/c++/v1/functional:1699:10
2020-01-11T03:45:23.3580425Z     #24 0x149c8c1f in std::__1::__function::__value_func<void ()>::~__value_func() /opt/llvm/bin/../include/c++/v1/functional:1821:19
2020-01-11T03:45:23.3580663Z     #25 0x149bf044 in std::__1::function<void ()>::~function() /opt/llvm/bin/../include/c++/v1/functional:2406:43
2020-01-11T03:45:23.3580939Z     #26 0x1d8ff730 in void std::__1::allocator_traits<std::__1::allocator<std::__1::__list_node<std::__1::function<void ()>, void*> > >::__destroy<std::__1::function<void ()> >(std::__1::integral_constant<bool, false>, std::__1::allocator<std::__1::__list_node<std::__1::function<void ()>, void*> >&, std::__1::function<void ()>*) /opt/llvm/bin/../include/c++/v1/memory:1747:23
2020-01-11T03:45:23.3581307Z     #27 0x1d8ff00f in void std::__1::allocator_traits<std::__1::allocator<std::__1::__list_node<std::__1::function<void ()>, void*> > >::destroy<std::__1::function<void ()> >(std::__1::allocator<std::__1::__list_node<std::__1::function<void ()>, void*> >&, std::__1::function<void ()>*) /opt/llvm/bin/../include/c++/v1/memory:1595:14
2020-01-11T03:45:23.3581613Z     #28 0x1d8fe649 in std::__1::__list_imp<std::__1::function<void ()>, std::__1::allocator<std::__1::function<void ()> > >::clear() /opt/llvm/bin/../include/c++/v1/list:762:13
2020-01-11T03:45:23.3581862Z     #29 0x1d8fe214 in std::__1::__list_imp<std::__1::function<void ()>, std::__1::allocator<std::__1::function<void ()> > >::~__list_imp() /opt/llvm/bin/../include/c++/v1/list:741:3
2020-01-11T03:45:23.3582119Z     #30 0x1d8fe1b7 in std::__1::list<std::__1::function<void ()>, std::__1::allocator<std::__1::function<void ()> > >::~list() /opt/llvm/bin/../include/c++/v1/list:834:28
2020-01-11T03:45:23.3582434Z     #31 0x213c4e15 in Envoy::Event::DispatcherImpl::~DispatcherImpl() /proc/self/cwd/source/common/event/dispatcher_impl.cc:57:1
2020-01-11T03:45:23.3582680Z     #32 0x213c4f64 in Envoy::Event::DispatcherImpl::~DispatcherImpl() /proc/self/cwd/source/common/event/dispatcher_impl.cc:53:35
2020-01-11T03:45:23.3582914Z     #33 0x1d730531 in std::__1::default_delete<Envoy::Event::Dispatcher>::operator()(Envoy::Event::Dispatcher*) const /opt/llvm/bin/../include/c++/v1/memory:2338:5
2020-01-11T03:45:23.3583163Z     #34 0x1d7301cc in std::__1::unique_ptr<Envoy::Event::Dispatcher, std::__1::default_delete<Envoy::Event::Dispatcher> >::reset(Envoy::Event::Dispatcher*) /opt/llvm/bin/../include/c++/v1/memory:2593:7
2020-01-11T03:45:23.3583430Z     #35 0x1d701158 in std::__1::unique_ptr<Envoy::Event::Dispatcher, std::__1::default_delete<Envoy::Event::Dispatcher> >::~unique_ptr() /opt/llvm/bin/../include/c++/v1/memory:2547:19
2020-01-11T03:45:23.3583676Z     #36 0x1d847599 in Envoy::Server::InstanceImpl::~InstanceImpl() /proc/self/cwd/source/server/server.cc:125:1
2020-01-11T03:45:23.3583897Z     #37 0x1d847984 in Envoy::Server::InstanceImpl::~InstanceImpl() /proc/self/cwd/source/server/server.cc:108:31
2020-01-11T03:45:23.3584118Z     #38 0x149582d1 in std::__1::default_delete<Envoy::Server::InstanceImpl>::operator()(Envoy::Server::InstanceImpl*) const /opt/llvm/bin/../include/c++/v1/memory:2338:5
2020-01-11T03:45:23.3584375Z     #39 0x14957f6c in std::__1::unique_ptr<Envoy::Server::InstanceImpl, std::__1::default_delete<Envoy::Server::InstanceImpl> >::reset(Envoy::Server::InstanceImpl*) /opt/llvm/bin/../include/c++/v1/memory:2593:7
2020-01-11T03:45:23.3584643Z     #40 0x14957b08 in std::__1::unique_ptr<Envoy::Server::InstanceImpl, std::__1::default_delete<Envoy::Server::InstanceImpl> >::~unique_ptr() /opt/llvm/bin/../include/c++/v1/memory:2547:19
2020-01-11T03:45:23.3585273Z     #41 0x1495776d in Envoy::MainCommonBase::~MainCommonBase() /proc/self/cwd/bazel-out/k8-dbg/bin/source/exe/_virtual_includes/envoy_main_common_lib/exe/main_common.h:35:7
2020-01-11T03:45:23.3586017Z     #42 0x14895fed in Envoy::MainCommon::~MainCommon() /proc/self/cwd/bazel-out/k8-dbg/bin/source/exe/_virtual_includes/envoy_main_common_lib/exe/main_common.h:94:7
2020-01-11T03:45:23.3586295Z     #43 0x149b55a5 in std::__1::default_delete<Envoy::MainCommon>::operator()(Envoy::MainCommon*) const /opt/llvm/bin/../include/c++/v1/memory:2338:5
2020-01-11T03:45:23.3586668Z     #44 0x149b532c in std::__1::unique_ptr<Envoy::MainCommon, std::__1::default_delete<Envoy::MainCommon> >::reset(Envoy::MainCommon*) /opt/llvm/bin/../include/c++/v1/memory:2593:7
2020-01-11T03:45:23.3587277Z     #45 0x149c740a in Envoy::AdminRequestTest::startEnvoy()::'lambda'()::operator()() const /proc/self/cwd/test/exe/main_common_test.cc:180:20
2020-01-11T03:45:23.3588022Z     #46 0x149c6771 in decltype(std::__1::forward<Envoy::AdminRequestTest::startEnvoy()::'lambda'()&>(fp)()) std::__1::__invoke<Envoy::AdminRequestTest::startEnvoy()::'lambda'()&>(Envoy::AdminRequestTest::startEnvoy()::'lambda'()&) /opt/llvm/bin/../include/c++/v1/type_traits:3530:1
2020-01-11T03:45:23.3588793Z     #47 0x149c65b1 in void std::__1::__invoke_void_return_wrapper<void>::__call<Envoy::AdminRequestTest::startEnvoy()::'lambda'()&>(Envoy::AdminRequestTest::startEnvoy()::'lambda'()&) /opt/llvm/bin/../include/c++/v1/__functional_base:348:9
2020-01-11T03:45:23.3589538Z     #48 0x149c64ff in std::__1::__function::__alloc_func<Envoy::AdminRequestTest::startEnvoy()::'lambda'(), std::__1::allocator<Envoy::AdminRequestTest::startEnvoy()::'lambda'()>, void ()>::operator()() /opt/llvm/bin/../include/c++/v1/functional:1533:16
2020-01-11T03:45:23.3590315Z     #49 0x149c0ea4 in std::__1::__function::__func<Envoy::AdminRequestTest::startEnvoy()::'lambda'(), std::__1::allocator<Envoy::AdminRequestTest::startEnvoy()::'lambda'()>, void ()>::operator()() /opt/llvm/bin/../include/c++/v1/functional:1707:12
2020-01-11T03:45:23.3590632Z     #50 0x14c1d139 in std::__1::__function::__value_func<void ()>::operator()() const /opt/llvm/bin/../include/c++/v1/functional:1860:16
2020-01-11T03:45:23.3590949Z     #51 0x14c1c09a in std::__1::function<void ()>::operator()() const /opt/llvm/bin/../include/c++/v1/functional:2419:12
2020-01-11T03:45:23.3591208Z     #52 0x270d5244 in Envoy::Thread::ThreadImplPosix::ThreadImplPosix(std::__1::function<void ()>)::$_0::operator()(void*) const /proc/self/cwd/source/common/common/posix/thread_impl.cc:33:9
2020-01-11T03:45:23.3591457Z     #53 0x270d50fc in Envoy::Thread::ThreadImplPosix::ThreadImplPosix(std::__1::function<void ()>)::$_0::__invoke(void*) /proc/self/cwd/source/common/common/posix/thread_impl.cc:32:7
2020-01-11T03:45:23.3591984Z     #54 0x7f7b49fd96b9 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76b9)
2020-01-11T03:45:23.3592496Z     #55 0x7f7b49af741c in clone (/lib/x86_64-linux-gnu/libc.so.6+0x10741c)
2020-01-11T03:45:23.3592694Z 
2020-01-11T03:45:23.3593244Z SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior source/server/listener_manager_impl.cc:695:61 in 
2020-01-11T03:45:23.3593489Z ================================================================================
2020-01-11T03:46:14.6870716Z [8,854 / 8,863] 537 / 545 tests, 1 failed; Linking test/integration/rtds_integration_test; 207s remote ... (8 actions running)
2020-01-11T03:47:41.0693020Z [8,861 / 8,865] 541 / 545 tests, 1 failed; Testing //test/integration:hotrestart_test; 121s remote ... (3 actions running)
2020-01-11T03:48:33.9377655Z INFO: Elapsed time: 689.961s, Critical Path: 589.11s
2020-01-11T03:48:33.9378899Z INFO: 6171 processes: 5858 remote cache hit, 1 local, 312 remote.
2020-01-11T03:48:34.5319968Z INFO: Build completed, 1 test FAILED, 8865 total actions
```

Comment (mattklein123) [reply source/server/listener_manager_impl.cc]: From a quick look I assume this is a test case in which the workers/listeners don't actually start up? I think that is what is happening. If so I think what you have is probably fine but can you double check this is the case and then add a comment about why we are not using Cleanup so that someone doesn't come along later and try to change it?

/wait

Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: Yes. That is the case. I added comment and test case reference. PTAL

File: source/server/listener_manager_impl.cc [hunk 5]
```diff
@@ -13,6 +13,7 @@
 #include "envoy/stats/scope.h"
 
 #include "common/common/assert.h"
+#include "common/common/cleanup.h"
```
Comment (mattklein123) [inline source/server/listener_manager_impl.cc]: del

File: source/server/listener_manager_impl.cc [hunk 6]
```diff
@@ -686,17 +691,30 @@ void ListenerManagerImpl::startWorkers(GuardDog& guard_dog) {
...
   uint32_t i = 0;
+
+  // We can not use "Cleanup" to simplify this logic here, because it results in a issue if Envoy is
+  // killed before workers are actually started. Specifically the AdminRequestGetStatsAndKill test
+  // case in main_common_test fails with ASAN error if we use "Cleanup" here.
+  const auto listeners_pending_init =
+      std::make_shared<std::atomic<uint64_t>>(workers_.size() * active_listeners_.size());
   for (const auto& worker : workers_) {
-    ENVOY_LOG(info, "starting worker {}", i);
+    ENVOY_LOG(info, "starting worker {}", i++);
```
Comment (mattklein123) [inline source/server/listener_manager_impl.cc]: Why this change? This seems like a behavior change below for the stats?

File: source/server/listener_manager_impl.cc (hunk 5 above)
Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: sorry, forgot to remove. Removed now.

File: source/server/listener_manager_impl.cc (hunk 6 above)
Comment (ramaraochavali) [reply source/server/listener_manager_impl.cc]: I have removed it for now. But I think it is not a behavioural change, but with the current code it logs "starting worker 0"  for all workers, which is confusing. I will do a separate PR for it. We can discuss there.
